policies/coding-style.md

```diff
@@ -586,11 +586,74 @@ it portable enough to meet our multi-platform support goals.
 
 ## Chapter 14:  Portability
 
+### Language
+
 To maximise portability the version of C defined in ISO/IEC 9899:1990
 should be used. This is more commonly referred to as C90. ISO/IEC 9899:1999
 (also known as C99) is not supported on some platforms that OpenSSL is
 used on and therefore should be avoided.
 
+### Integer types
+
+The types `int8_t`, `uint8_t`, `int16_t`, `uint16_t`, `int32_t`, `uint32_t`,
+`int64_t` and `uint64_t` are recommended for use when the exact width is
+important. These were only standardised in C99 but we define them ourselves if
+they are not available on a target platform.
+
+`size_t` is available. `ossl_ssize_t` should be used instead of `ssize_t` for
+portability.
+
+Most of the built-in integer types in the C language should be avoided
+or have important caveats:
+
+- Use of `long` and `unsigned long` is deprecated, as these types cannot be
+  relied upon to be larger than `int` on all platforms.
+
+  Use `size_t` to represent array lengths and indices. If a signed value is
+  needed, use `ossl_ssize_t`.
+
+- `short` or `unsigned short` should not be used in public APIs or to denote
+  a width. These types should be used only where the range of values required is
+  a subset of the range `[-32767, 32767]`, as this range is guaranteed by the C
+  standard.
+
+- `int` and `unsigned int` can be used.
+
+  The C standard only guarantees that `int` can represent the range
+  `[-32767, 32767]`, but it is an assumption of the OpenSSL codebase that `int`
+  is at least 32 bits in size.
+
+  When using `unsigned int`, write `unsigned int` instead of just `unsigned`.
+
+- `long long` and `unsigned long long` are deprecated; use `int64_t` or
+  `uint64_t` instead.
+
+- Use `char` for integers only where the range of values required is a subset of
+  the range `[0, 127]`, as this range is guaranteed by the C standard. (`char`
+  is unsigned by default on some platforms.)
+
+  However, consider using `int8_t` instead.
+
+- Use `signed char` only where the range of values required is a subset of
+  the range `[-127, 127]`, as this range is guaranteed by the C standard.
+
+  As with `char`, consider using `int8_t` instead.
+
+- Use `unsigned char` for byte buffers. `unsigned char` may also be
+  used for integers, but consider using `uint8_t` instead.
+
+- Use `char` for strings, but bear in mind that `char` is unsigned by default
+  on some platforms.
```

Comment:
Or should we define that all strings in public APIs are in UTF8?

Reply:
Note that C11 allows `u8"string"` and that represents an UTF8 string stored as unsigned char. We should probably use unsigned char strings and deprecate char.

I  understand that a future C standard (C2X) will introduce a char8_t that is a distinct type that works with u8.

Reply:
We don't use `wchar_t` anywhere, I don't think it bears talking about.

String encoding is an attribute of the public API, not coding style, so I don't think it belongs here.

I don't think using unsigned char strings is realistic or desirable. Some platforms have signed char, other platforms have unsigned char, in either case those are what OS APIs expect to receive. It would be an overbearing change to the codebase, and should be discussed in a separate proposal.

Reply:
This policy is not about how we talk to the OS. This is how applications talk to the library. We currently do not say anything about encoding of characters. As a result, you can't write applications that properly display the strings. Our own apps just write something to the console and hope the user can read it. One way of dealing with that is is saying that it needs to be UTF8 encoded. An other way of dealing with it is using something like wchar_t.

You can argue that string encoding is not a coding style, but you're now placing it in our coding style.

We should not be directly writing strings to the OS, so I don't see a problem with it not being a char.

Reply:
This PR doesn't say anything about string encoding.

Reply:
And that is the problem. This does not improve anything. All new APIs will still have the same problem that it has some unknown encoding.

Reply:
If we want to have a technical policy on string encoding that's fine, but that is completely orthogonal to integer types and belongs in a separate PR.

I see limited scope for a policy on string encoding since we interact with the OS using functions like `fopen` etc. and thus essentially inherit the OS's string encoding. We could have a policy which basically says nothing but "our policy is the OS's policy" but material improvements seem unavailable. "Everything must be UTF-8" is a non-starter.